presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java

```diff
@@ -357,57 +344,30 @@ public void moreMemoryAvailable()
-        driverContext.transferMemoryToTaskContext(bytes);
-
-        TaskContext taskContext = driverContext.getPipelineContext().getTaskContext();
-        if (taskBytes > bytes) {
-            try {
-                taskContext.reserveMemory(taskBytes - bytes);
-            }
-            catch (ExceededMemoryLimitException e) {
-                taskContext.freeMemory(bytes);
-                throw e;
-            }
-        }
-        else {
-            taskContext.freeMemory(bytes - taskBytes);
-        }
-    }
-
-    public void setMemoryReservation(long newMemoryReservation)
-    {
-        checkArgument(newMemoryReservation >= 0, "newMemoryReservation is negative");
-
-        long delta = newMemoryReservation - memoryReservation.get();
-
-        if (delta > 0) {
-            reserveMemory(delta);
-        }
-        else {
-            freeMemory(-delta);
-        }
-    }
-
-    public boolean trySetMemoryReservation(long newMemoryReservation)
+    /**
+     * This method is <b>not</b> thread safe. If multiple threads work on the same
+     * <code>transferredBytesMemoryContext</code> instance it's possible that one of the writes
+     * is lost by the last <code>transferredBytesMemoryContext.setBytes()</code> call.
+     * @param taskBytes The number of bytes to transfer from the OperatorContext to TaskContext.
+     * @param transferredBytesMemoryContext This context should be created with {@link com.facebook.presto.operator.TaskContext#createNewTransferredBytesMemoryContext()}.
+     */
+    public void transferMemoryToTaskContext(long taskBytes, LocalMemoryContext transferredBytesMemoryContext)
     {
-        checkArgument(newMemoryReservation >= 0, "newMemoryReservation is negative");
-
-        long delta = newMemoryReservation - memoryReservation.get();
-
-        if (delta > 0) {
-            if (!driverContext.tryReserveMemory(delta)) {
-                return false;
-            }
-
-            memoryReservation.addAndGet(delta);
+        long bytes = operatorMemoryContext.getUserMemory();
+        operatorMemoryContext.localUserMemoryContext().transferOwnership(transferredBytesMemoryContext);
+        try {
+            transferredBytesMemoryContext.setBytes(transferredBytesMemoryContext.getBytes() + taskBytes - bytes);
         }
-        else {
-            freeMemory(-delta);
+        catch (ExceededMemoryLimitException e) {
+            // Since we can reserve an extra amount of memory above (when taskBytes > bytes),
+            // ExceededMemoryLimitException can be thrown when reserving this extra amount.
+            // This will happen after "bytes" is moved from the operator context to the task context.
+            // At this point "bytes" isn't tracked by the operator context.
+            // When drivers get destroyed, only memory tracked in its operators are freed.
+            // Therefore, we have to cleanup the memory here.
+            transferredBytesMemoryContext.setBytes(0);
```

Comment:
For this, you you would need to verify that `transferredBytesMemoryContext.getBytes()` was initially `0`.

All in all, this looks like coding to the exact (and only) implementation of `LocalMemoryContext.transferMemory`, which means that general-purpose-looking `LocalMemoryContext` remains implicitly coupled with `OperatorContext`.

Reply:
I updated this to reset the state of `transferredBytesMemoryContext` to its previous state before the transfer.

Reply:
Good. Did you also add some synchronization so that we know the state wasn't changed in the meantime?